### PR TITLE
Fix crashes in expression evaluation. Add compatibility fixes.

### PR DESCRIPTION
**Pull request by allenss-amazon**
Crash fix (always-on): APPLY of any string function (lower, upper, strlen,
  startswith, contains, substr, concat, timefmt, parsetime) on a nil-resolving
  operand aborted the server. No longer.

  Behavior changes (selected by search.emulate-release; each exposed as an
  INFO search_compatibility counter): non-empty strings are now truthy in
  &&/||/!; lower/upper return nil for non-string args; abs/ceil/exp/floor/log/
  log2/sqrt return NaN for non-numeric strings; month() rounds correctly to
  the first of the month (was off by one day); parsetime returns nil on
  format mismatch; timefmt with empty format returns nil; date functions
  return nil for pre-epoch and ±inf/NaN timestamps.

  Added integration/compatibility/generate_expr.py for verification.

### CODE REVIEW

File: src/expr/value.cc [hunk 1]
```diff
@@ -30,6 +31,11 @@ static bool IsNan(const double& d) {
   return ((v & kExponentMask) == kExponentMask) && ((v & kMantissaMask) != 0);
 }
 
+static bool IsInf(const double& d) {
+  uint64_t v = *(uint64_t*)&d;
+  return ((v & kExponentMask) == kExponentMask) && ((v & kMantissaMask) == 0);
+}
```
[the remark this answers is not in the document]
Comment (allenss-amazon) [reply src/expr/value.cc:38]: These are fine.

File: src/valkey_search_options.cc [hunk 2]
```diff
@@ -743,5 +707,41 @@ config::Number& GetMutationWeightTag() {
...
+  }
+  return absl::OkStatus();
+}
+
+static auto emulate_release_config =
+    config::VersionBuilder(kEmulateReleaseConfig,
+                           vmsdk::ValkeyVersion(kModuleVersion.Major(), 0, 0),
+                           kEmulateReleaseMin, kEmulateReleaseMaxInfinity)
+        .WithValidationCallback(ValidateEmulateRelease)
+        .Build();
```
[the remark this answers is not in the document]
Comment (allenss-amazon) [reply src/valkey_search_options.cc:753]: this is fine.

File: vmsdk/src/module_config.cc [hunk 3]
```diff
@@ -346,14 +341,87 @@ absl::Status String::FromString(std::string_view value) {
...
+  if (!absl::SimpleAtod(text, &value)) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("Invalid double value: '", text, "'"));
+  }
+  return value;
+}
+
+std::string ConfigTraits<double>::Format(double value) {
+  return absl::StrFormat("%g", value);
+}
```
[the remark this answers is not in the document]
Comment (allenss-amazon) [reply vmsdk/src/module_config.cc:362]: this is fine.

File: testing/expr/value_test.cc
Comment (Frank-Gu-81) [inline testing/expr/value_test.cc:1]: I think integration tests are enough, but still wanted to check if we want to add a few cheap unit tests as well since it looks like the core crash fix (string funcs on a nil-resolving operand returning Nil instead of the old CHECK(false)) and the non-finite date hardening are only checked via the integration sweep. 

Kiro suggested:

```
TEST_F(ValueTest, StringFuncsOnNilReturnNil) {
   Value nil;  // previously aborted via CHECK(false)
   EXPECT_TRUE(FuncStrlen(nil).IsNil());
   EXPECT_TRUE(FuncLower(nil).IsNil());
   EXPECT_TRUE(FuncUpper(nil).IsNil());
   EXPECT_TRUE(FuncConcat({nil}).IsNil());
   EXPECT_TRUE(FuncStartswith(nil, Value("x")).IsNil());
   EXPECT_TRUE(FuncContains(nil, Value("x")).IsNil());
   EXPECT_TRUE(FuncSubstr(nil, Value(0.0), Value(1.0)).IsNil());
 }

 TEST_F(ValueTest, DateFuncsOnNonFiniteReturnNil) {
   Value nan_v(std::nan(""));
   for (const Value& bad : {pos_inf, neg_inf, nan_v}) {
     EXPECT_TRUE(FuncYear(bad).IsNil());
     EXPECT_TRUE(FuncMonth(bad).IsNil());
     EXPECT_TRUE(FuncDayofweek(bad).IsNil());
     EXPECT_TRUE(FuncHour(bad).IsNil());
   }
 }
```

**Review by Frank-Gu-81 — APPROVED** [review]
